### PR TITLE
marketplace example

### PR DESCRIPTION
**Pull request by machisuji**
First try at a simple marketplace dApp.

image

The screenshot above is the result of running the demo which is doing the following:

```
$XRD_MARKET = resim call-function $PACKAGE Market open $XRD

resim call-method $XRD_MARKET buy $USDT 3.21 1000,$XRD
resim call-method $XRD_MARKET buy $USDT 3.27 200,$XRD
resim call-method $XRD_MARKET sell 90,$USDT 3.26
resim call-method $XRD_MARKET sell 100,$USDT 3.29
resim call-method $XRD_MARKET sell 400,$USDT 3.33

resim call-method $XRD_MARKET fill_orders

resim call-method $XRD_MARKET print_order_book
```

Still have to write the actual tests beyond the demo.

### CODE REVIEW

**Comment by machisuji** [comment]
> Does someone have an idea why this (https://github.com/machisuji/community-scrypto/blob/example/marketplace/defi/marketplace/src/lib.rs#L147) isn't working? Perhaps I can't have vaults in structs? Each order has two vaults here (purchase and payment). Maybe this all isn't working like I imagine?

Ok it's because of the ticket which isn't burned, stored or returned. I'll have to look into how I can make it so it can be burned.

**Comment by machisuji** [comment]
Withdrawal works correctly too now.

**Comment by machisuji** [comment]
I know there's a lot of duplication there between sales and purchases which can probably be handled all by exactly the same code. Wanted to get it working first. Might refactor this so that the struct, and code for placing and withdrawing orders is shared.

**Comment by Aos-dev** [comment]
A very interesting example. Thank you for thinking about both your Windows and Linux colleagues :)

Possible improvement: I believe the market does not take into account the sequence in which the orders are placed. The transactional price is always the seller's price what puts the sellers at a disadvantage. The easiest way to solve it would be probably to fill the orders as soon as a new order is placed (providing there is something to fill) but that would be also a change in the overall behaviour.

Once again, a very educative example and with nice potential to expand it further: stop orders, stop limit orders, market orders. Do you think about any of those?

**Comment by machisuji** [comment]
> A very interesting example. Thank you for thinking about both your Windows and Linux colleagues :)
> 
> Possible improvement: I believe the market does not take into account the sequence in which the orders are placed. The transactional price is always the seller's price what puts the sellers at a disadvantage. The easiest way to solve it would be probably to fill the orders as soon as a new order is placed (providing there is something to fill) but that would be also a change in the overall behaviour.
> 
> Once again, a very educative example and with nice potential to expand it further: stop orders, stop limit orders, market orders. Do you think about any of those?

Thanks for the comment. Yeah I was wondering how real markets do this so that the seller isn't always at a disadvantage. Even when filling the orders right away what would happen if there's a sell order for 0.26 and someone, not paying attention, puts in a buy order for 0.27. Would they just pay 0.27 or would they get the cheaper price of the available sell order?

I was thinking about implementing all those different kinds of orders, yeah! Have to read up a bit more about that before, though.  My financial experience is actually quite limited!

**Comment by Aos-dev** [comment]
Regarding your example the person should get the cheaper price of the available sell order.
The price you enter when placing an order (i.e. the limit pirice) is the maximum price you are willing to pay, but of course you will be even more happy and accept a lower one (and symmetrically when placing a sell order).
In a bit more complex example lets say there are 3 sell orders: 100@0.26, 100@0.27 and 100@0.28. Now somebody enters a buy order 250@0.27. This order would immediately "consume" 100@0.26 (paying 0.26 for a unit because why shoud you pay more :)), 100@0.27 and there will be a leftover buy of 50@0.27. When now somebody enters a sell of 100@0.25 this will "consume" this lefover (but at the price of 0.27 because why should you sell cheaper :)) and leave its own leftover of a sell of 50@0.25. And so on...

**Comment by machisuji** [comment]
Ah thanks for the explanation. I'll adapt the code so that orders are simply always filled (or tried to be filled) right away when a new one is placed.

**Comment by machisuji** [comment]
> So you see, the easiest would be to fill the orders "on the fly" every time a new one is placed. If not, you have to remember the sequence in which the orders entered the book. The one at a disadvantage is always the older one.

I'm filling the orders on-the-fly as you suggested now and now it behaves correctly, I think. The demo (.ps1) demonstrates that. Have yet to update the bash demo accordingly.

I've got market and limit orders for now. Will implement stop and stop limit orders next.

File: defi/marketplace/src/data.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        self.token.metadata()["symbol"].clone()
+    }
+
+    pub fn currency(&self) -> String {
+        self.payment.resource_def().metadata()["symbol"].clone()
+    }
+
+    pub fn is_filled(&self) -> bool {
+        if self.buy {
+            self.payment.amount() == 0.into() || self.payment.amount() < self.price
```
Comment (iamyulong) [inline defi/marketplace/src/data.rs:34]: I'd be interested to know if we can buy a fraction of a token.

File: defi/marketplace/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+                .metadata("symbol", format!("OT-{}", order_number))
+                .metadata("order_number", format!("{}", order_number))
+                .metadata("order_token_address", token_address.to_string())
+                .metadata("order_currency", self.base_currency())
+                .new_token_mutable(self.ticket_minter_badge.resource_def());
+
+            self.ticket_minter_badge.authorize(|badge|{
+                (ticket_resource_def.mint(1, badge), ticket_resource_def.address())
+            })
+        }
```
Comment (iamyulong) [inline defi/marketplace/src/lib.rs:48]: With Scrypto v0.2.0, we can now use NFT to represent a ticket without creating too many resources.

**Comment by iamyulong** [comment]
Very good demonstration of order-book based marketplace. I'm going to merge it now, but feel free to make further PRs for enhancements.

**Comment by machisuji** [comment]
Thanks @iamyulong! I'll definitely create new PRs with enhancements. Especially using NFTs for the tickets makes a lot of sense.

**Pull request merged**